File: sendkey/public/assets/crypto.js
// SendKey envelope v1, browser implementation.
// Must stay byte-for-byte compatible with crypto.go:
//
//   outer ct = AES-256-GCM(K, outerIV, envelope)
//   envelope = version(0x01) || flags(1) || body
//   flags bit0 set: body = salt(16) || innerIV(12) || innerCT
//     innerKey = PBKDF2-HMAC-SHA256(passphrase, salt, 310000, 32 bytes)
//
// K is generated here, carried only in the URL fragment, and never sent
// anywhere: the server stores ciphertext it cannot read.

const VERSION = 0x01;
const FLAG_PASSPHRASE = 0x01;
const FLAG_FILE = 0x02; // body is a file manifest, not the secret itself
const SALT_LEN = 16;
const IV_LEN = 12;
const PBKDF2_ITERS = 310000;

export const b64u = {
  encode(bytes) {
    let s = '';
    for (const b of bytes) s += String.fromCharCode(b);
    return btoa(s).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
  },
  decode(str) {
    const s = atob(str.replace(/-/g, '+').replace(/_/g, '/'));
    const out = new Uint8Array(s.length);
    for (let i = 0; i < s.length; i++) out[i] = s.charCodeAt(i);
    return out;
  },
};

function randomBytes(n) {
  const b = new Uint8Array(n);
  crypto.getRandomValues(b);
  return b;
}

function concat(...parts) {
  const out = new Uint8Array(parts.reduce((n, p) => n + p.length, 0));
  let off = 0;
  for (const p of parts) { out.set(p, off); off += p.length; }
  return out;
}

async function aesKey(rawBytes, usages) {
  return crypto.subtle.importKey('raw', rawBytes, 'AES-GCM', false, usages);
}

async function deriveInnerKey(passphrase, salt, usages) {
  const material = await crypto.subtle.importKey(
    'raw', new TextEncoder().encode(passphrase), 'PBKDF2', false, ['deriveKey']);
  return crypto.subtle.deriveKey(
    { name: 'PBKDF2', salt, iterations: PBKDF2_ITERS, hash: 'SHA-256' },
    material, { name: 'AES-GCM', length: 256 }, false, usages);
}

// seal encrypts secretBytes; returns { ct, iv, key } as Uint8Arrays.
// extraFlags adds envelope flags (FLAG_FILE for manifests); the passphrase
// bit is managed here and stripped from extraFlags.
export async function seal(secretBytes, passphrase, extraFlags = 0) {
  let flags = extraFlags & ~FLAG_PASSPHRASE;
  let body = secretBytes;

  if (passphrase) {
    flags |= FLAG_PASSPHRASE;
    const salt = randomBytes(SALT_LEN);
    const innerIV = randomBytes(IV_LEN);
    const innerKey = await deriveInnerKey(passphrase, salt, ['encrypt']);
    const innerCT = new Uint8Array(
      await crypto.subtle.encrypt({ name: 'AES-GCM', iv: innerIV }, innerKey, secretBytes));
    body = concat(salt, innerIV, innerCT);
  }

  const envelope = concat(new Uint8Array([VERSION, flags]), body);
  const keyBytes = randomBytes(32);
  const iv = randomBytes(IV_LEN);
  const key = await aesKey(keyBytes, ['encrypt']);
  const ct = new Uint8Array(
    await crypto.subtle.encrypt({ name: 'AES-GCM', iv }, key, envelope));

  return { ct, iv, key: keyBytes };
}

// openOuter peels the outer layer only. Returns { flags, body } or throws
// Error('bad-key') when K is wrong / data corrupted.
export async function openOuter(ctBytes, ivBytes, keyBytes) {
  const key = await aesKey(keyBytes, ['decrypt']);
  let env;
  try {
    env = new Uint8Array(
      await crypto.subtle.decrypt({ name: 'AES-GCM', iv: ivBytes }, key, ctBytes));
  } catch {
    throw new Error('bad-key');
  }
  if (env.length < 2 || env[0] !== VERSION) throw new Error('bad-version');
  return { flags: env[1], body: env.subarray(2) };
}

export function needsPassphrase(flags) {
  return (flags & FLAG_PASSPHRASE) !== 0;
}

export function isFile(flags) {
  return (flags & FLAG_FILE) !== 0;
}

// openInner decrypts a passphrase-protected body. Throws
// Error('bad-passphrase') on a wrong passphrase. Retryable locally, the
// server is never contacted again.
export async function openInner(body, passphrase) {
  if (body.length < SALT_LEN + IV_LEN + 1) throw new Error('malformed');
  const salt = body.subarray(0, SALT_LEN);
  const innerIV = body.subarray(SALT_LEN, SALT_LEN + IV_LEN);
  const innerCT = body.subarray(SALT_LEN + IV_LEN);
  const innerKey = await deriveInnerKey(passphrase, salt, ['decrypt']);
  try {
    return new Uint8Array(
      await crypto.subtle.decrypt({ name: 'AES-GCM', iv: innerIV }, innerKey, innerCT));
  } catch {
    throw new Error('bad-passphrase');
  }
}

File: sendkey/crypto.go
package sendkey

// SendKey envelope format v1
// --------------------------
// This is the wire format shared byte-for-byte between the Go CLI and the
// browser (public/assets/crypto.js). The server never sees anything below the
// outer layer.
//
//	outer ciphertext = AES-256-GCM(K, outerIV, envelope)
//	  K       : 32 random bytes, carried ONLY in the URL fragment (#...)
//	  outerIV : 12 random bytes, stored with the ciphertext (not secret)
//
//	envelope = version(1) || flags(1) || body
//	  version : 0x01
//	  flags   : bit0 = passphrase-protected
//
//	body (flags&1 == 0):  secret bytes
//	body (flags&1 == 1):  salt(16) || innerIV(12) || innerCT
//	  innerKey = PBKDF2-HMAC-SHA256(passphrase, salt, 310000, 32)
//	  innerCT  = AES-256-GCM(innerKey, innerIV, secret bytes)
//
// The passphrase layer sits INSIDE the outer layer, so a recipient who typed
// the wrong passphrase can retry locally, forever, without touching the
// server — the one burn-on-read fetch already happened.

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"errors"
	"fmt"
)

const (
	envVersion      = 0x01
	flagPassphrase  = 0x01
	flagFile        = 0x02 // body is a file manifest, not the secret itself
	saltLen         = 16
	IVLen           = 12      // outer/inner GCM nonce length
	KeyLen          = 32      // AES-256 key length
	pbkdf2Iters     = 310_000 // OWASP recommendation for PBKDF2-HMAC-SHA256
	maxEnvelopeSize = 1 << 20 // sanity bound when opening
)

// ErrBadPassphrase is returned when the inner layer fails to authenticate.
var ErrBadPassphrase = errors.New("wrong passphrase")

// Sealed is an envelope encrypted under a fresh random key.
type Sealed struct {
	CT  []byte // outer ciphertext for the server
	IV  []byte // outer nonce for the server
	Key []byte // 32-byte key for the URL fragment — never send this anywhere
}

// Seal encrypts secret into a v1 envelope. If passphrase is non-empty an
// inner PBKDF2-derived layer is added.
func Seal(secret []byte, passphrase string) (*Sealed, error) {
	return SealWithFlags(secret, passphrase, 0)
}

// SealWithFlags is Seal with extra envelope flags, used for file manifests
// (FlagFile). The passphrase bit is managed here and must not be passed in.
func SealWithFlags(secret []byte, passphrase string, extra byte) (*Sealed, error) {
	body := secret
	flags := extra &^ flagPassphrase

	if passphrase != "" {
		salt := make([]byte, saltLen)
		innerIV := make([]byte, IVLen)
		if err := fillRandom(salt, innerIV); err != nil {
			return nil, err
		}
		innerKey := pbkdf2Key([]byte(passphrase), salt, pbkdf2Iters, KeyLen)
		innerCT, err := gcmSeal(innerKey, innerIV, secret)
		if err != nil {
			return nil, err
		}
		flags |= flagPassphrase
		body = make([]byte, 0, saltLen+IVLen+len(innerCT))
		body = append(body, salt...)
		body = append(body, innerIV...)
		body = append(body, innerCT...)
	}

	envelope := make([]byte, 0, 2+len(body))
	envelope = append(envelope, envVersion, flags)
	envelope = append(envelope, body...)

	key := make([]byte, KeyLen)
	outerIV := make([]byte, IVLen)
	if err := fillRandom(key, outerIV); err != nil {
		return nil, err
	}
	ct, err := gcmSeal(key, outerIV, envelope)
	if err != nil {
		return nil, err
	}
	return &Sealed{CT: ct, IV: outerIV, Key: key}, nil
}

// FlagFile is the envelope flag marking the body as a file manifest.
const FlagFile = flagFile

// NeedsPassphrase opens only the outer layer and reports whether the envelope
// carries an inner passphrase layer.
func NeedsPassphrase(ct, iv, key []byte) (bool, error) {
	env, err := openOuter(ct, iv, key)
	if err != nil {
		return false, err
	}
	return env[1]&flagPassphrase != 0, nil
}

// EnvelopeFlags opens only the outer layer and returns the raw flags byte.
func EnvelopeFlags(ct, iv, key []byte) (byte, error) {
	env, err := openOuter(ct, iv, key)
	if err != nil {
		return 0, err
	}
	return env[1], nil
}

// Open decrypts a v1 envelope. passphrase may be empty when the envelope has
// no inner layer; a wrong passphrase yields ErrBadPassphrase (retryable —
// the outer ciphertext is already in hand, nothing else burns).
func Open(ct, iv, key []byte, passphrase string) ([]byte, error) {
	env, err := openOuter(ct, iv, key)
	if err != nil {
		return nil, err
	}
	flags, body := env[1], env[2:]

	if flags&flagPassphrase == 0 {
		return body, nil
	}
	if len(body) < saltLen+IVLen+1 {
		return nil, errors.New("malformed passphrase envelope")
	}
	salt, innerIV, innerCT := body[:saltLen], body[saltLen:saltLen+IVLen], body[saltLen+IVLen:]
	innerKey := pbkdf2Key([]byte(passphrase), salt, pbkdf2Iters, KeyLen)
	secret, err := gcmOpen(innerKey, innerIV, innerCT)
	if err != nil {
		return nil, ErrBadPassphrase
	}
	return secret, nil
}

func openOuter(ct, iv, key []byte) ([]byte, error) {
	if len(key) != KeyLen {
		return nil, fmt.Errorf("key must be %d bytes", KeyLen)
	}
	if len(ct) > maxEnvelopeSize {
		return nil, errors.New("envelope too large")
	}
	env, err := gcmOpen(key, iv, ct)
	if err != nil {
		return nil, errors.New("decryption failed: wrong key or corrupted data")
	}
	if len(env) < 2 || env[0] != envVersion {
		return nil, errors.New("unsupported envelope version")
	}
	return env, nil
}

func gcmSeal(key, iv, plaintext []byte) ([]byte, error) {
	g, err := newGCM(key)
	if err != nil {
		return nil, err
	}
	return g.Seal(nil, iv, plaintext, nil), nil
}

func gcmOpen(key, iv, ct []byte) ([]byte, error) {
	g, err := newGCM(key)
	if err != nil {
		return nil, err
	}
	return g.Open(nil, iv, ct, nil)
}

func newGCM(key []byte) (cipher.AEAD, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	return cipher.NewGCM(block)
}

func fillRandom(bufs ...[]byte) error {
	for _, b := range bufs {
		if _, err := rand.Read(b); err != nil {
			return err
		}
	}
	return nil
}
